Label systemd configuration files with systemd_conf_t

zpytela opened the pull request:
The systemd_conf_t type was added as default file context for files with the .conf suffix and .conf.d directories in /etc/systemd, /run/systemd, and /usr/lib/systemd.
The /usr/local/lib/systemd directory is covered by file equivalency. The systemd_domain attribute was allowed read access to these files.

Refer to https://github.com/systemd/systemd/blob/main/NEWS CHANGES WITH 256-rc1:
General Changes and New Features:

        * Various programs will now attempt to load the main configuration file
          from locations below /usr/lib/, /usr/local/lib/, and /run/, not just
          below /etc/. For example, systemd-logind will look for
          /etc/systemd/logind.conf, /run/systemd/logind.conf,
          /usr/local/lib/systemd/logind.conf, and /usr/lib/systemd/logind.conf,
          and use the first file that is found.  This means that the search
          logic for the main config file and for drop-ins is now the same.

Resolves: rhbz#2279923

martinpitt commented:
Interesting -- this breaks `systemctl disable --now dnf-automatic dnf-automatic-install`, which feels a little too specific to this PR to be a coincidence. @mvollmer @jelly can you please take over? I'm on my way out to PTO, sorry.

zpytela commented:
This PR was expected to have unpredictable impact, but I don't see any relation to dnf neither I can reproduce it. What I can confirm is some mess regarding dnf4 and dnf5 plugins in rawhide.

If you find any particular problem, please point me to it. I am not going to merge this PR right away.

AdamWill commented:
@martinpitt we landed the move from dnf 4 to dnf 5 by default in rawhide yesterday. it seems more likely that that would cause the problem than this PR?

martinpitt commented:
@AdamWill Ah! That would be it then, dnf5-automatic doesn't have the -install service/timer variant any more. Thanks!

zpytela committed:
Label systemd configuration files with systemd_conf_t

The systemd_conf_t type was added as default file context for plain
files with the .conf suffix, for .conf.d directories in /etc/systemd,
/run/systemd, and /usr/lib/systemd, and for plain files and symlinks
in those directories. The /usr/local/lib/systemd directory is a subject
of file equivalency rules.
The systemd_domain attribute was allowed read access to these files.

Refer to https://github.com/systemd/systemd/blob/main/NEWS
CHANGES WITH 256-rc1:
General Changes and New Features:

        * Various programs will now attempt to load the main configuration file
          from locations below /usr/lib/, /usr/local/lib/, and /run/, not just
          below /etc/. For example, systemd-logind will look for
          /etc/systemd/logind.conf, /run/systemd/logind.conf,
          /usr/local/lib/systemd/logind.conf, and /usr/lib/systemd/logind.conf,
          and use the first file that is found.  This means that the search
          logic for the main config file and for drop-ins is now the same.

Resolves: rhbz#2279923

zpytela committed:
Allow systemd-hostnamed read the vsock device

The commit addresses the following AVC denial:
type=PROCTITLE msg=audit(05/10/2024 14:57:37.213:1388) : proctitle=/usr/lib/systemd/systemd-hostnamed
type=PATH msg=audit(05/10/2024 14:57:37.213:1388) : item=0 name=/dev/vsock inode=346 dev=00:06 mode=character,666 ouid=root ogid=root rdev=0a:7a obj=system_u:object_r:vsock_device_t:s0 nametype=NORMAL cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=SYSCALL msg=audit(05/10/2024 14:57:37.213:1388) : arch=x86_64 syscall=openat success=no exit=EACCES(Permission denied) a0=AT_FDCWD a1=0x7fc898b3e5be a2=O_RDONLY|O_CLOEXEC a3=0x0 items=1 ppid=1 pid=17773 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=systemd-hostnam exe=/usr/lib/systemd/systemd-hostnamed subj=system_u:system_r:systemd_hostnamed_t:s0 key=(null)
type=AVC msg=audit(05/10/2024 14:57:37.213:1388) : avc:  denied  { open } for  pid=17773 comm=systemd-hostnam path=/dev/vsock dev="devtmpfs" ino=346 scontext=system_u:system_r:systemd_hostnamed_t:s0 tcontext=system_u:object_r:vsock_device_t:s0 tclass=chr_file permissive=0

zpytela committed:
Label /run/systemd/machine with systemd_machined_var_run_t

Along with that, a transition for systemd-machined for this
directory was defined.

The commit addresses the following AVC denial:
type=PROCTITLE msg=audit(05/10/2024 12:11:33.651:224) : proctitle=/usr/lib/systemd/systemd-machined
type=PATH msg=audit(05/10/2024 12:11:33.651:224) : item=1 name=/run/systemd/machine nametype=CREATE cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=PATH msg=audit(05/10/2024 12:11:33.651:224) : item=0 name=/run/systemd/ inode=2 dev=00:1a mode=dir,755 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:init_var_run_t:s0 nametype=PARENT cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=SYSCALL msg=audit(05/10/2024 12:11:33.651:224) : arch=x86_64 syscall=mkdirat success=no exit=EACCES(Permission denied) a0=AT_FDCWD a1=0x5635d5a3cc21 a2=0755 a3=0x0 items=2 ppid=1 pid=4216 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=systemd-machine exe=/usr/lib/systemd/systemd-machined subj=system_u:system_r:systemd_machined_t:s0 key=(null)
type=AVC msg=audit(05/10/2024 12:11:33.651:224) : avc:  denied  { create } for  pid=4216 comm=systemd-machine name=machine scontext=system_u:system_r:systemd_machined_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=dir permissive=0

Resolves: rhbz#2279923

keszybz commented on policy/modules/system/systemd.fc:
```diff
@@ -4,6 +4,8 @@ HOME_DIR/\.config/systemd/user(/.*)?		gen_context(system_u:object_r:systemd_unit
 /etc/.*hostname.*	--	gen_context(system_u:object_r:hostname_etc_t,s0)
 /etc/machine-info	--	gen_context(system_u:object_r:hostname_etc_t,s0)
+/etc/systemd/.*\.conf	--	gen_context(system_u:object_r:systemd_conf_t,s0)
+/etc/systemd/.*\.conf\.d(/.*)?	gen_context(system_u:object_r:systemd_conf_t,s0)
```
I guess this could be changed to `/etc/systemd/.*\.conf\.d(/.*\.conf)?`, but it doesn't matter much.

zpytela replied:
Can be, although it may be another punch to clarity. Can there be only plain files?
```
/etc/systemd/.+\.conf	--	gen_context(system_u:object_r:systemd_conf_t,s0)
/etc/systemd/.+\.conf\.d	-d	gen_context(system_u:object_r:systemd_conf_t,s0)
/etc/systemd/.+\.conf\.d/.+\.conf	--	gen_context(system_u:object_r:systemd_conf_t,s0)
```

keszybz replied:
This version is nice.

> Can there be only plain files?

Only `*.conf` is read. They don't have to be "plain", a symlink would be accepted too by the systemd code.

zpytela replied:
OK, updated. The level of complexity now looks like an overkill.
I will merge it later tomorrow. thanks for the review.

keszybz reviewed (approved):
Looks all reasonable.
